### DISCUSSION TITLE
Alternative syntax for actions

### DISCUSSION THREAD

**Discussion by konnov** [informalsystems/quint#207] [Ideas]
As we have discussed recently, the syntax for action-level operators `<-`, `guess`, `&`, and `|` seems to be less intuitive than the syntax for other operators. For instance, here is one alternative for `guess`: https://github.com/informalsystems/tnt/discussions/183#discussioncomment-3579203

At some point, we would like to conduct a user survey, to see which syntax is more intuitive to the engineers, who have never seen TLA+/Quint before. Let's collect all potential alternatives here. No judgement, all ideas are welcome. But no guarantee that any of these alternatives would be implemented.

Here are the four operators in question and their current syntax in Quint/TLA+:

1. Assignment: `x <- e` in Quint and `x' = e` in TLA+.
2. Non-deterministic choice: `S.guess({ x => e })` in Quint and `\E x \in S: e` in TLA+.
3. Action-level conjunction: `{ A1 & ... & An }` in Quint and `A1 /\ ... /\ An` in TLA+.
4. Action-level disjunction: `{ A1 | ... | An }` in Quint and `A1 \/ ... \/ An` in TLA+.

**Comment by konnov** [comment]
One idea for conjunctions and disjunctions is to make them look a bit more like JS promises (but not exactly like them):

1. Action-level conjunction: `all { A1, ..., An }`.
2. Action-level disjunction: `oneof { A1, ..., An }`. JS has `any`, but I don't like it :)

**Reply by thpani** [reply]
Looks neat! WDYT about `branch` instead of `oneof`?

**Reply by konnov** [reply]
`branch` sounds like `if-then-else` to me :)

**Reply by thpani** [reply]
Makes sense.
People might think of `oneof` as implying depth-first search; `any` makes more sense to me.

**Reply by thpani** [reply]
Actually, we already use `S.guess` for data non-determinism.
So why not `guess { A1, ... , An }` for control non-determinism?

Or maybe simply `nondet { A1, ..., An }`.

**Reply by konnov** [reply]
> So why not guess { A1, ... , An } for control non-determinism?

We are avoiding operator overloading. So technically, we should not have two operators with the same name.

**Reply by thpani** [reply]
I thought of it more as a keyword.
In any case, I'd expect them to be distinguishable on the syntax level?

**Reply by shonfeder** [reply]
How about `choice { .... }`?

**Reply by thpani** [reply]
`choice` is very close to `CHOOSE`, which could be confusing wrt (non)determinism.

**Reply by shonfeder** [reply]
Oh yeah, good point.

**Reply by konnov** [reply]
As a provisional solution I will replace `&` with `all` and `|` with `any`, as `&` and `|` seem to annoy people. Let's continue the discussion in the hope that we find even a better solution.

**Reply by shonfeder** [reply]
So to be clear for my view: `&` and `|` do not annoy me. i like them. What annoys me is not being able to use them uniformly.

**Comment by konnov** [comment]
More discussions about `guess`: https://github.com/informalsystems/tnt/pull/216#discussion_r978214664

**Comment by konnov** [comment]
I have been thinking about `<-` and `guess`, and Shon's suggestion. How about the following:

 - We introduce the special syntax `x <- oneOf(S)` that is a syntax shorthand for `S.guess(y => x <- e)`.
 - We allow the users to write `next(x)` in actions.

As a result, in simple cases we will have a combination of `oneOf` and `next`, e.g.:

```scala
initializer Init = all {
  x <- oneOf(Int),
  next(x) % 2 == 0
}
```

In more complex cases, the users can still use `guess`, as it is the most general form. For instance, the above example could be written as:

```
initializer Init =
  Int.guess(i => all {
    i % 2 == 0,
    x <- i
  })
```

**Comment by shonfeder** [comment]
# How should we represent the relation between states?

## 0. Context

A number of related questions have been circulating in our discussions:

- What syntax should we provide for specifying initial states?
  https://github.com/informalsystems/tnt/discussions/183#discussioncomment-3579203
- What syntax should we provide for specifying non-deterministic constraints on
  (state) variable "assignments"?
  https://github.com/informalsystems/tnt/discussions/183#discussioncomment-3579203
- What syntax will provide ergonomic and precise specification of variables that
  do not change between states?
  https://github.com/informalsystems/tnt/discussions/70
- What syntax (if any) should we use to refer to the value of a variable in the
  next state? https://github.com/informalsystems/tnt/discussions/207#discussioncomment-3774581
- This discussion, asking which sorts of syntax sugar we should use for actions
generally.

More recently, in Quint meetings, we've pondered these questions:

- What syntax should we use for supplying values or constraints for parametric
  rigid variables (aka "CONSTANTS")?
- What do we do about the likely confusion resulting from our current
  "assignment" syntax?

The root issue in all these questions seems to be instability or inadequacy in
the ways we can currently represent the relation between states. In this thread,
I'll try to help spin up some discussion around this more holistic consideration.

My current hypothesis is that most of the problems we are seeing arise form lack of a clear and perspicuous method of representing the relations (qua relational) between states, which is the core concern of an actions. As per Lamport: "An action represents a relation between old states and new states". In the replies and discussion, I hope to tease out this hypothesis a bit.

**Reply by shonfeder** [reply]
## 1. "Delayed assignment" -- `<-`
 
### Anticipated problem

We have discussed a usability concern with `<-`. The concern can be illustrated
with the following example:

```scala
action Foo = all {
  x == 2,
  x <- 1,
  x == 1,
}
```

This transition predicate will always be false, but we are pretty sure engineers
are going to expect it to be true, based on the combination of two factors:

1. The common understanding of the semantics of assignment and the relatively common use of `<-` as [notation for assignment in PLs](https://en.wikipedia.org/wiki/Assignment_(computer_science)#Notation) and the absolutely common use of it to mean assignment in [pseudocode](https://en.wikipedia.org/wiki/Pseudocode#Common_mathematical_symbols).
2. The tendency to read a sequence of lines within `{...}` as specifying an
   order of execution.
   
Either of these alone would cause confusion on their own, but they also
interfere.

In `tnt/doc/lang.md` we introduce `<-` as "Delayed assignment", but it's
unlikely like that the "delayed" qualifier or our explanation that "TNT
assignment is delayed by one step" in any docs is going to prevent engineers
from writing specs like

```scala
x <- e
...
foo(x)
```

with the expectation that `x` will be `e` *now*. Engineers reading specs are
liable to make the same mistake, and this is a recipe for specification and/or
implementation errors.

`<-` was proposed to avoid the misconceptions observed by people encountering
the `'` state function in TLA+. It could be useful to recall what those misconceptions were.  I'm worried that, as is, our solution may introduce problems that are at least as bad.

@konnov mentioned he was partially inspired by Go's use of `<-` for
communication via channels, so that `x <- e` could be read as "send `e` to `x`
in the next state"). This makes sense, but two key difference in Go's channel
syntax show why the current approach is error prone:

- In Go, channels are a distinct type. You can send and receive values on a
  channel, but cannot use the channel in the place where you'd expect the value,
  which means you can't make the same mistakes illustrated above.
- In Go, `<-` is used symmetrically to send and receive messages from channels.
  There are two different syntactic constructs for
  [sending](https://go.dev/ref/spec#Send_statements) and [receiving](
  https://go.dev/ref/spec#Receive_operator) from channels, and this statically
  enforces sensible interaction with channels.
  
Our current design enjoys neither of these protections. 

A bit of user research could help test this concern.

#### As related to relations

Here's an annotated translation of our example:

```scala
// The relation between two states, `s` and `t` called `Foo`
action Foo = all { \\ holds if all of the following are true
  x == 2,  \\ x in state s is 2
  x <- 1,  \\ x is assigned to 1 in state t
  x == 1,  \\ x in state s is 1
}
```

Once we have made the state relation clear **as a relation between states** it is clear that the relation cannot hold, since `x` cannot be both `2` and `1` in the same state. However, the current syntax doesn't make this relationality apparent, due to, at least:

1. the baggage associated with `<-`
2. lacking differentiation between `x` at `s` and `x` at `t`

### Some alternatives

So far, 2 broad categories of alternatives have occurred to me:

#### A. Something like priming

Embrace something like TLA+ syntax, that allows clear reference to the vars in this and the next state. 

In TLA+, variables are "state functions", and the state to which a variable is to be applied is left implicit. The priming function `'` is introduced to manage this implicit function application: `f'` meanings "apply the state function `f` to the *new* state, rather the current one". In the special case of variables: 

> the unprimed variables refer to the old state and the primed variables refer to the new state 
> -- *The Temporal Logic of Actions*, p. 4

I suspect that *part* of the problem users run into when trying to think via TLA+ syntax is due to this implicitness. However, IMO, falling back to priming `'` in the limited context of variables would be better than our current situation, as it at least gives an indication of the two different states being related.

However, I'm particularly interested in approaches that would make the relation explicit and @bugarela shared some ideas and interest in this kind of approach. iirc, it was also used in an experimental demo developed by Roman and Andrey and others.

##### Reified states

Explicitly passing the states as 1st class parameters seems honest, since transition predicates are relations between states. This amounts to desugaring the `A` out of `TLA`, and giving the predicates back their hidden arguments:

```scala
def Foo(s0: state, s1: state) = all {
  s0.x == 2,
  s1.x <- 1, // could double as update syntax for records
  s0.x == 1,
}
```

This could also make the difference between state predicates and initializers apparent, without the need for keywords:

- State predicates would only take `s0`
- Init predicates would only take `s1`

We might also consider how our row-typed records would enable static analysis on partitions of the state-space with the state reified into records in this way.

This approach essentially follows the characterization of a state as a function `Var -> Val`, from variables to states. However, we'd need a record rather a map due to the typed nature of our variant: each variable must be able to have a different type, which indicates that we are dealing with a product type, viz. a record.

Having to explicit thread the state through every action would get tedious, so we'd probably end up with something akin to [Prolog's DCGs](https://en.wikipedia.org/wiki/Definite_clause_grammar), which would make the states more implicit again, which is probably more or less exactly how Lamport landed on the "actions macros system" in the first place :)

##### Explicit states via keywords

Rather than passing states explicitly as params to operators, we could use keywords akin to `this` in Scala:

```scala
action Foo = all {
  now.x == 2,
  next.x <- 1,
  now.x == 1,
}
```

where `now` and `next` (or `old/new` or `s0/s1` or `this/next` or whatever) are ambient, but scoped within a particular action.

This brings us very close to the `next` operator already defined for temporal predicates. A variant of this approach is to use keywords as qualifiers or operators rather than as special name spaces:

```scala
action Foo = all {
  x == 2,
  next x <- 1, // equivalently next(x) <- 1
  x == 1,
}
```

Making the current state implicit we lose the very clear and unambiguous differentiation between `x`, and I think there's still some chance that readers would think the `x` on the next **line** should be updated "next", but this at least seems error prone.

With `next x <- 1` and `1 == next x` we may be a little closer to Go channels, with `next _` being like a wrapper that protects us from thinking we are reading or updating the variable in the current state.

#### B. Embrace the more functional/operational approach

I suspect confusion here does and will arise from the kind of awkward balance struck between operational idioms and the underlying relational meaning of the transition predicates. Another way to solve this mismatch is to bring the syntax farther away from the relational specification.

This might be achieved by only allowing constraints on the next-state variables to occur "at the end" of an action. This would amount to some sort of precondition-then-update idiom like:

```scala
action Foo = when all {
  x == 2,
  x == 1,
} then {
  x <- 1
}
```

This kind of thing has been discussed already, and was suggested by Andrey and others.

**Reply by konnov** [reply]
That's a very good analysis @shonfeder! We should definitely keep thinking about better syntax for assignments, which should be understandable to engineers.

From a short-term perspective, it would be good to converge on a new syntax for the non-deterministic choice, which is currently implemented via `guess`. As we discussed, the main issue with the current syntax for `guess` is that it hides assignments inside its lambda argument, e.g.:

```scala
Set(1, 2, 3).guess(i => x <- i)
```

So the only way to refer to the next value of `x` is by using `i` inside the lambda body. Alternatively, we could have a special syntax form for non-deterministic choice that follows the `val` construct:

```scala
input val i = guess(Set(1, 2, 3))
and {
  i > 0,
  x <- i
}
```

What I like about this approach is that it uses the same pattern as `val` and `pure val`. However, we should enforce (via the effects type checker?) that `guess` only appears on the right-hand side of `input val`.

I don't have strong feelings about the keyword `input`. We could use `nondet` or another keyword.

**Reply by shonfeder** [reply]
Could you remind me what problem the `input` keyword is meant to solve? I'm a bit unclear on that. I.e., why couldn't we just allow use of val like

```scala
val i = guess(Set(1,2,3))
...
```

**Reply by bugarela** [reply]
I'm only reading this right now (sorry) and my first impression is that I love this alternative:
```
action Foo = all {
  now.x == 2,
  next.x <- 1,
  now.x == 1,
}
```

It's probably the less strict one we ever had for TNT, and it's still better then the prime operator, which can be applied to arbitrary expressions. I think it would help even the most experienced spec writers to read a specification and it makes actions actually look like transition specification, which was a think Zarko was missing when we talked about TNT at the retreat. I really really like this.

**Reply by shonfeder** [reply]
This kind of approach also resonates with me! I'd like to experiment a bit with how it might help or contribute to addressing the other questions enumerated, to see if there might be teeth to it :)

**Reply by konnov** [reply]
> Could you remind me what problem the `input` keyword is meant to solve? I'm a bit unclear on that. I.e., why couldn't we just allow use of val like
> 
> ```scala
> val i = guess(Set(1,2,3))
> ...
> ```

The problem with `val` is that it can be used anywhere in the code. For instance, if we let `guess` inside function constructors, we would get the non-deterministic behavior in seemingly deterministic expressions, similar to what we had in https://github.com/informalsystems/apalache/issues/841#issuecomment-980003708

My idea was allow for this kind of special value only in actions, which should be checkable with effects.

**Reply by konnov** [reply]
Another idea for the guessing qualifier:

```
oracle val i = guess(Set(1, 2, 3))
```

I actually like the `oracle` qualifier the most. An oracle can be implemented in various ways: as non-deterministic choice, as random choice, as user input. The keyword should not suggest one mechanism.

**Reply by konnov** [reply]
> I'm only reading this right now (sorry) and my first impression is that I love this alternative:
> 
> ```
> action Foo = all {
>   now.x == 2,
>   next.x <- 1,
>   now.x == 1,
> }
> ```
> 
> It's probably the less strict one we ever had for Quint, and it's still better then the prime operator, which can be applied to arbitrary expressions. I think it would help even the most experienced spec writers to read a specification and it makes actions actually look like transition specification, which was a think Zarko was missing when we talked about Quint at the retreat. I really really like this.

I agree that this approach looks much more self-explanatory, though `now` and `next` are kind of implicit global objects, that are not variables. Dafny has the operator `old(x)` to refer to the variable value upon function entry: https://dafny.org/latest/DafnyRef/DafnyRef#sec-old-expression. So instead of `x` and `next(x)`, they have `old(x)` and `x`. I have always thought that it was weird, but it probably makes a lot of sense from the engineers perspective.

How about the following. We already have `next(x)`, so we can write assignments like:

```scala
next(x) <- e
```

We could introduce `now(x)` which is simply an identity operator. We could say that this operator could be omitted, but using it improves readability.

So the above example would look like:

```scala
action Foo = all {
  now(x) == 2,
  next(x) <- 1,
  now(x) == 1,
}
```

I agree that it would improve the spec clarity a lot, for the price of nothing :)

**Reply by konnov** [reply]
Another option. We introduce `old(x)` instead of `now(x)` and recommend using `old(x)`, once `next(x) <- e` was written. So the example would look like follows:

```scala
action Foo = all {
  x == 2,
  next(x) <- 1,
  old(x) == 1,
}
```

By doing so, we are saving on typing `old(x)`, when it is not necessary, and also give the user means to disambiguate what `x` refers to.

**Reply by shonfeder** [reply]
At the moment, the syntax I like best for relating states is:

1. Adopt the proposal for lexical indication of state variables from #374 
2. Use primes for "next state", but don't introduce them as a more general operator shifting the semantics of state functions.
3. Ditch `<-`, and expect to teach people how to think about relations between constraints on states.

Our running example would then look like this:

```ruby
action Foo = all {
  @x  == 2,
  @x' == 1,
  @x  == 1,
}
```

**Reply by konnov** [reply]
Nice, I think we are converging to something nice here. I agree that we could introduce primes back, if we just treat a prime as a special assignment syntax, not an individual operator (as in TLA+). For instance:

```
action Foo = all {
  x == 2,
  x' = 1,
  x == 1,
}
```

Note that I would like to distinguish between assignments with `=` and equalities `==`. Although an assignment is indeed an equality constraint (as in TLA+), assignments are more than that: they are unconditionally true, they modify `x'`, and they can do only one modification in a transition.

**Comment by konnov** [comment]
Also closing this one, as we are not going to change the related syntax in the near future.
